src/core/frames/frame_controller.ts

```diff
@@ -172,14 +172,15 @@ export class FrameController
     this.reloadable = false
     this.formSubmission = new FormSubmission(this, element, submitter)
     const { fetchRequest } = this.formSubmission
-    this.prepareHeadersForRequest(fetchRequest.headers, fetchRequest)
+    const id = getAttribute("data-turbo-frame", submitter, element) || this.element.getAttribute("target")
+    this.prepareHeadersForRequest(fetchRequest.headers, fetchRequest, id)
     this.formSubmission.start()
   }
 
   // Fetch request delegate
 
-  prepareHeadersForRequest(headers: FetchRequestHeaders, _request: FetchRequest) {
-    headers["Turbo-Frame"] = this.id
+  prepareHeadersForRequest(headers: FetchRequestHeaders, _request: FetchRequest, id?: string | null) {
+    headers["Turbo-Frame"] = id || this.id
   }
 
   requestStarted(_request: FetchRequest) {
```

Comment:
I suggest using the existing method `this.findFrameElement` to locate the turbo-frame.  Perhaps something like:

```diff
  formSubmissionIntercepted(element: HTMLFormElement, submitter?: HTMLElement) {
    if (this.formSubmission) {
      this.formSubmission.stop()
    }
  
    this.reloadable = false
    this.formSubmission = new FormSubmission(this, element, submitter)
    const { fetchRequest } = this.formSubmission
-   this.prepareHeadersForRequest(fetchRequest.headers, fetchRequest)
+   const frame = this.findFrameElement(element, submitter)
+   this.prepareHeadersForRequest(fetchRequest.headers, fetchRequest, frame);
    this.formSubmission.start()
  }

// Fetch request delegate

- prepareHeadersForRequest(headers: FetchRequestHeaders, _request: FetchRequest) {
+ prepareHeadersForRequest(headers: FetchRequestHeaders, _request: FetchRequest, frame: FrameElement = this.element) {
-   headers["Turbo-Frame"] = this.id
+   headers["Turbo-Frame"] = frame.id
  }
```

Reply:
This was my first approach but... if the target frame is missing, -- which throws an error but doesn't interrupt the form submission -- it won't send the correct header.

Reply:
The diff I show above does exactly what you suggest.  Do have a sample where it's not working?

Reply:
````html
<turbo-frame id="frame">
  <form data-turbo-frame="hello">
    (...)
  </form>
</turbo-frame>
````

`findFrameElement` always return a FrameElement.
If there is no `<turbo-frame id="hello">` in the current page, it will return the parent frame `<turbo-frame id="frame">` instead.

For reference:
````js
private findFrameElement(element: Element, submitter?: HTMLElement) {
  const id = getAttribute("data-turbo-frame", submitter, element) || this.element.getAttribute("target")
  return getFrameElementById(id) ?? this.element
}
````

Given so, the form submission will send a wrong header `Turbo-Frame: frame` if the target frame (hello) is missing.

With the solution I provided, we just try to extract the target ID to send the correct header, nevermind the target frame is present or not. Throwing an exception is another concern.

However, we can refactor the ID extraction into a new function `findFrameId`.
It's also used in [shouldInterceptNavigation](https://github.com/hotwired/turbo/blob/main/src/core/frames/frame_controller.ts#L333)


````diff
   formSubmissionIntercepted(element: HTMLFormElement, submitter?: HTMLElement) {
     [...]
-    this.prepareHeadersForRequest(fetchRequest.headers, fetchRequest)
+    const id = findFrameId(element, submitter)
+    this.prepareHeadersForRequest(fetchRequest.headers, fetchRequest, id)
     this.formSubmission.start()
   }

+  private findFrameId(element: Element, submitter?: HTMLElement) {
+    const id = getAttribute("data-turbo-frame", submitter, element) || this.element.getAttribute("target")
+    return id ?? this.id
+  }

   private findFrameElement(element: Element, submitter?: HTMLElement) {
-    return getFrameElementById(id) ?? this.element
+    const id = getAttribute("data-turbo-frame", submitter, element) || this.element.getAttribute("target")
+    return getFrameElementById(findFrameId(element, submitter)
   }

[...]

   private shouldInterceptNavigation(element: Element, submitter?: HTMLElement) {
-    const id = getAttribute("data-turbo-frame", submitter, element) || this.element.getAttribute("target")
+    const id = findFrameId(element, submitter)
````

Reply:
I've pushed two more commits :
- [one test to verify header sent when target frame is missing](https://github.com/hotwired/turbo/pull/579/commits/9a3a3654e1e06cad6c58adda856388fcb7428111)
- [refactor of ID extraction](https://github.com/hotwired/turbo/pull/579/commits/7c4657375f294ab36fe6171155d8948734375d0f)

Reply:
The original implementation of `data-turbo-frame` by @seanpdoyle purposely falls back to parent turbo-frame.  The comment from the commit states:

> When navigating via an `<a>` click within a `<turbo-frame>`, determine the target of the navigation by checking for `[data-turbo-frame]` overrides on the `<a>` element itself, and then check for frames referenced by `turbo-frame[target]`, then fall back to the frame itself.

see: https://github.com/hotwired/turbo/commit/f62ad3b3069597c60334b1d0623c53d6432252ff

The proposed changes creates inconsistent behavior for links and form submissions when the target for `data-turbo-frame` is missing.

- link: falls back to parent turbo-frame and sends request
- form: fails request

Reply:
You're right, it should be the same behavior for links and forms.
I also agree with behavior that fall back to the parent frame to performs DOM updates.

But IMHO, the header we sent should be the frame it was meant to be, no matter the frame exists or not.

I understand that this may be open to debate and may lead to breaking changes, so I'll refactor this PR to use `findFrameElement` and open another issue to debate discuss this last behavior.